### DISCUSSION TITLE
move custom definition of the %__spec_check_pre from spec file to global macros causes not the same result

### DISCUSSION THREAD

**Discussion by kloczek** [rpm-software-management/rpm#2864] [Q&A]
**Describe the bug**
in https://github.com/rpm-software-management/rpm/issues/316#issuecomment-1900766922 I wrote that I'm using +3 years solution which allows me to disable %check execution by use `%bcond_with check`.
What I'm using is not exactly what I've proposed in next comment.
What I'm using is purely spec files solution by add in spec file like below
```spec
%bcond_with check
.
.
%check
%{!?with_check:echo "%%check disabled."; exit 0}
<test suite_execution>
```

<details>
<summary>What wrote in the in #316 was based on below test case:</summary>

```spec
Summary:        bcond check test
Name:           test
Version:        1
Release:        1
License:        foo

%description
bcond test.

%prep

%define        __spec_check_pre \
       %{!?with_check:echo "%%check disabled."; exit 0} \
       %{___build_pre}

%check
echo "%{name} package test suite"

%files
```

```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bi test.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.lNjy8k
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.mQr3fh
%check disabled.
Processing files: test-1-1.x86_64
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/tkloczko/rpmbuild/BUILDROOT/test-1-1.x86_64
```
</details>

However when I've moved `%__spec_check_pre` definition to macros.check  file
```spec
# Add handle --{with|without} check
%__spec_check_pre \
        %{!?with_check:echo "%%check disabled."; exit 0} \
        %{___build_pre}
```
and started testing spec file like below
```spec
Summary:        bcond check test
Name:           test
Version:        1
Release:        1
License:        foo

%description
bcond test.

%prep

%check
echo "%{name} package test suite"

%files
```
Result is not the same:
```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bi test.spec --macros /usr/lib/rpm/macros:/usr/lib/rpm/redhat/macros:macros.check
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.bnQsY0
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.JZfaOi
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ echo 'test package test suite'
test package test suite
+ RPM_EC=0
++ jobs -p
+ exit 0
Processing files: test-1-1.x86_64
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/tkloczko/rpmbuild/BUILDROOT/test-1-1.%{_arch}
```
Looks like it is not possible to evaluate in bcond macros from global macros or that evaluation needs to be added differently.

Q: what I'm doing incorrectly and/or how to eval `%with_chec` from global macros? 🤔

**Comment by kloczek** [comment]
Sorry Actual behaviour is different.
When I've executed above commands I've already had changed `%__spec_check_pre` in /usr/lib/rpm/redhat/macros.

<details>
<summary>Result with defined %bcond_with</summary>

```spec
%bcond_with     check

Summary:        bcond test
Name:           test
Version:        1
Release:        1
License:        foo

%description
bcond test.

%prep

# Add handle --{with|without} check
%define __spec_check_pre \
        %{!?with_check:echo "%%check disabled."; exit 0} \
        %{___build_pre}

%check
echo "%{name} package test suite"
exit 1

%files
```

```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bi test.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.4hXN9D
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.XMbSqT
%check disabled.
Processing files: test-1-1.x86_64
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/tkloczko/rpmbuild/BUILDROOT/test-1-1.x86_64
```
</details>

Which is as expected

<details>
<summary>Result with defined %bcondout_with</summary>

```spec
%bcond_without  check

Summary:        bcond test
Name:           test
Version:        1
Release:        1
License:        foo

%description
bcond test.

%prep

# Add handle --{with|without} check
%define __spec_check_pre \
        %{!?with_check:echo "%%check disabled."; exit 0} \
        %{___build_pre}

%check
echo "%{name} package test suite"
exit 1

%files
```

```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bi test.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.wJqF1l
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.pKu6HS
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ echo 'test package test suite'
test package test suite
+ exit 1
error: Bad exit status from /var/tmp/rpm-tmp.pKu6HS (%check)
```
</details>

Which is as expected as well.

<details>
<summary>Result without defined bcond</summary>

```spec
Summary:        bcond test
Name:           test
Version:        1
Release:        1
License:        foo

%description
bcond test.

%prep

# Add handle --{with|without} check
%define __spec_check_pre \
        %{!?with_check:echo "%%check disabled."; exit 0} \
        %{___build_pre}

%check
echo "%{name} package test suite"
exit 1

%files
```

```console
[tkloczko@pers-jacek SPECS]$ rpmbuild -bi test.spec
warning: source_date_epoch_from_changelog set but %changelog is missing
Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.KZbK71
+ umask 022
+ cd /home/tkloczko/rpmbuild/BUILD
+ RPM_EC=0
++ jobs -p
+ exit 0
Executing(%check): /bin/sh -e /var/tmp/rpm-tmp.6PTPuy
%check disabled.
Processing files: test-1-1.x86_64
Checking for unpackaged file(s): /usr/lib/rpm/check-files /home/tkloczko/rpmbuild/BUILDROOT/test-1-1.x86_64
```
</details>

Incorrect result because without define `%bcond_{with|without} check` such solution should not be disabling %check.

In other words it is not issue with %__spec_check_pre redefinition but above approach with globally redefines %__spec_check_pre does not work as expected (which is not rpm issue).

Q: how to modify above to have expected behaviour? 🤔

**Reply by dmnks** [reply]
The `%with_*` macro is a feature of `%bcond`. It won't ever be defined if `%bcond` isn't.

The answer to your question therefore is: make sure to have a `%bcond` definition in your SPEC. More info here: https://rpm-software-management.github.io/rpm/manual/conditionalbuilds.html
